### PR TITLE
Add logical_not operator.

### LINKED ISSUE #23836: explicit logical operators
We should have logical operators equivalent to NumPy:

https://docs.scipy.org/doc/numpy/reference/generated/numpy.logical_not.html
https://docs.scipy.org/doc/numpy/reference/generated/numpy.logical_and.html
https://docs.scipy.org/doc/numpy/reference/generated/numpy.logical_or.html
https://docs.scipy.org/doc/numpy/reference/generated/numpy.logical_xor.html

Most critical are `logical_not` and `logical_xor`, because are current error messages for bool are strange because we only have bitwise_not (which sort of implies our bools are 1-bit, but that's not correct).  For example:

```
>>> np.array(True) - np.array(False)
TypeError: numpy boolean subtract, the `-` operator, is deprecated, use the bitwise_xor, the `^` operator, or the logical_xor function instead.

>>> -np.array(False)
TypeError: The numpy boolean negative, the `-` operator, is not supported, use the `~` operator or the logical_not function instead.

>>> torch.tensor(True) - torch.tensor(False)
RuntimeError: Subtraction, the `-` operator, with two bool tensors is not supported. Use the `^` operator instead.

>>> -torch.tensor(False)
RuntimeError: Negation, the `-` operator, on a bool tensor is not supported. 
                        If you are trying to invert a mask, use the `~` or `bitwise_not()` operator instead."
```

### PR DESCRIPTION
**Pull request by xuhdev**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24379 Implement operator logical_and
* #23978 Let logical_xor support non-bool tensors.
* #23916 Let logical_not support non-bool tensors.
* #23860 Recommend logical_not() instead of bitwise_not() when applying sub and neg on bool tensors.
* #23847 Add logical_xor operator
* **#23839 Add logical_not operator.**

Close #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23839

Differential Revision: [D16678301](https://our.internmc.facebook.com/intern/diff/D16678301)

### CODE REVIEW

File: aten/src/ATen/Dispatch.h [hunk 1]
```diff
@@ -220,6 +220,23 @@ inline void deprecated_AT_DISPATCH_ALL_TYPES_AND_HALF_AND_COMPLEX() {}
     }                                                                        \
   }()
 
+#define AT_DISPATCH_INTEGRAL_TYPES_AND_BOOL(TYPE, NAME, ...)                 \
```
Comment (gchanan) [inline aten/src/ATen/Dispatch.h]: we've shied away from these crazily-named dispatch macros -- the ones that continue to be in here are here for BC reasons.

Can we call make a `AT_DISPATCH_INTEGRAL_TYPES_AND` macro and make it follow `AT_DISPATCH_ALL_TYPES_AND` (i.e. don't call `::detail::scalar_type`)?

**Review by gchanan — CHANGES_REQUESTED** [review]

File: torch/_torch_docs.py [hunk 2]
```diff
@@ -2536,6 +2536,32 @@ def merge_dicts(*dicts):
 
 """)
 
+add_docstr(torch.logical_not,
+           r"""
+logical_not(input, out=None) -> Tensor
+
+Computes the element-wise logical NOT of the given input tensor. The output tensor always has the same dtype as that of
```
Comment (gchanan) [inline torch/_torch_docs.py]: this is different than numpy's logical_not, which defaults to returning bool -- is there any reason why we don't do that?  Basically, I think this should behave the same way as our comparison operators.

File: torch/_torch_docs.py [hunk 3]
```diff
@@ -2536,6 +2536,32 @@ def merge_dicts(*dicts):
...
+:attr:`input`.
+
+- The logical NOT of a bool is the one that is the same as standard textbook definition: False -> True, True ->
```
Comment (gchanan) [inline torch/_torch_docs.py]: I don't really think you need this section, the examples are sufficient.

File: test/test_torch.py [hunk 4]
```diff
@@ -1819,6 +1819,25 @@ def _test_bitwise_not(self, device):
     def test_bitwise_not(self):
         self._test_bitwise_not(self, 'cpu')
 
+    @staticmethod
+    def _test_logical_not(self, device):
+        for dtype in (torch.bool, torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64,
```
Comment (gchanan) [inline test/test_torch.py]: can you use https://github.com/pytorch/pytorch/blob/9588cd921efd051a982188220b8215914bd5790e/torch/testing/__init__.py#L94-L102?

File: aten/src/ATen/Dispatch.h (hunk 1 above)
Comment (xuhdev) [reply aten/src/ATen/Dispatch.h]: I just realized that this is actually not used in this PR. I have removed this macro.

**Review by gchanan — CHANGES_REQUESTED** [review]

File: torch/_torch_docs.py (hunk 2 above)
Comment (xuhdev) [reply torch/_torch_docs.py]: If we do this, can we still have inplace operator? What I'm thinking is that, in this case, a user should use `bool()` to do the conversion if a user needs bool tensors. This is also consistent with what `bitwise_not()` currently does.

Comment (gchanan) [reply torch/_torch_docs.py]: I think it should be behave like a comparison ops; in that case inplace keeps the dtype:
```
>>> x.gt_(1)
tensor([[0., 0., 1.],
        [0., 0., 0.]])
```

btw if this is difficult it seems fine to just support boolean for now (so we can get it into 1.2) and then add the other dtypes in a follow-on patch.

Comment (xuhdev) [reply torch/_torch_docs.py]: OK, I'll change it to bool only for now so we can safely have it into v1.2.

File: aten/src/ATen/native/UnaryOps.cpp [hunk 5]
```diff
@@ -56,6 +56,27 @@ Tensor& bitwise_not_out(Tensor& result, const Tensor& self) {
...
+  return at::logical_not_out(result, self);
+}
+
+Tensor& logical_not_(Tensor& self) {
+  return at::logical_not_out(self, self);
+}
+
+Tensor& logical_not_out(Tensor& result, const Tensor& self) {
+  TORCH_CHECK(self.scalar_type() == kBool,
+              "logical_not only supports bool tensors.");
```
Comment (gchanan) [inline aten/src/ATen/native/UnaryOps.cpp]: nit: I'd say "only currently supports" to make it clear we are planning to add support for other types.

File: test/test_torch.py [hunk 6]
```diff
@@ -1819,6 +1819,24 @@ def _test_bitwise_not(self, device):
...
+            a = torch.tensor([10, 1, 0], dtype=dtype, device=device)
+            # new tensor
+            self.assertEqual(expected_res, a.logical_not())
+            # out
+            b = torch.empty(0, dtype=dtype, device=device)
+            torch.logical_not(a, out=b)
+            self.assertEqual(expected_res, b)
+            # in-place
+            a.logical_not_()
+            self.assertEqual(expected_res, a)
```
Comment (gchanan) [inline test/test_torch.py]: what happens if you pass the wrong dtype as an `out` parameter?

**Review by gchanan — APPROVED** [review]

**Comment by gchanan** [comment]
clang tidy failure looks bogus.

**Comment by gchanan** [comment]
@xuhdev: are all your recent changes to this stack just merges/rebases?  I need to know if landing a prior version is safe.

also, can you stop merging/rebasing?  I already checked that the failures weren't related to this diff, and we have some internal systems that will reject commits if the version we are trying to merge isn't up-to-date with the the PR.  I'm not sure if the merges will go through now or I'll have to do redo it.

I realize we have no automated communication on this state so this is totally invisible to you, sorry about that!

**Comment by xuhdev** [comment]
@gchanan Yes, I just rebased this morning, in case there's any conflict. Sorry for the confusion!

**Comment by gchanan** [comment]
ok, I think it makes the most sense to punt on getting this into v1.2 -- mainly because we can't land anything right now.

**Comment by gchanan** [comment]
Also, no need to apologize -- thanks so much for your contributions, they are really great.

**Comment by xuhdev** [comment]
@gchanan No problem! Thanks for your review!

**Pull request closed without merging**